File: cli-ios/RemoteCoding/Networking/AuthAPI.swift
import Foundation

/// 登录 REST(服务器 8080)。Google idToken → 本系统 {account, token}。
enum AuthAPI {
    struct AuthResult: Decodable {
        let account: String
        let token: String
        let name: String?
        let hasPassword: String?   // Google 返回:"true"/"false"(是否已设密码)
    }

    enum AuthError: LocalizedError {
        case server(String)
        var errorDescription: String? {
            if case .server(let m) = self { return m }
            return nil
        }
    }

    /// REST 基地址:沿用设置页的服务器 IP,端口固定 8080(Spring MVC)。
    @MainActor
    static var baseURL: String { "http://\(RelayClient.savedHost):8080/api/auth" }

    /// 认领电脑端的配对码:把 VibeNotch 绑到当前登录账号下。
    @MainActor
    static func claimPair(code: String) async throws {
        guard let token = AppState.sessionToken else { throw AuthError.server("请先登录") }
        guard let url = URL(string: "http://\(RelayClient.savedHost):8080/api/pair/claim") else {
            throw URLError(.badURL)
        }
        var req = URLRequest(url: url, timeoutInterval: 10)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.httpBody = try JSONEncoder().encode(["code": code, "token": token])
        let (data, resp) = try await URLSession.shared.data(for: req)
        guard let http = resp as? HTTPURLResponse else { throw URLError(.badServerResponse) }
        guard http.statusCode == 200 else {
            let msg = (try? JSONDecoder().decode([String: String].self, from: data))?["error"]
            throw AuthError.server(msg ?? "配对失败(\(http.statusCode))")
        }
    }

    /// 用 Google idToken 换本系统会话令牌。
    @MainActor
    static func loginWithGoogle(idToken: String) async throws -> AuthResult {
        guard let url = URL(string: "\(baseURL)/google") else { throw URLError(.badURL) }
        var req = URLRequest(url: url, timeoutInterval: 12)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.httpBody = try JSONEncoder().encode(["idToken": idToken])
        let (data, resp) = try await URLSession.shared.data(for: req)
        guard let http = resp as? HTTPURLResponse else { throw URLError(.badServerResponse) }
        guard http.statusCode == 200 else {
            let msg = (try? JSONDecoder().decode([String: String].self, from: data))?["error"]
            throw AuthError.server(msg ?? "登录失败(\(http.statusCode))")
        }
        return try JSONDecoder().decode(AuthResult.self, from: data)
    }

    /// 邮箱密码登录(日常主力:不依赖 Google,只连你自己的服务器)。
    @MainActor
    static func login(account: String, password: String) async throws -> AuthResult {
        guard let url = URL(string: "\(baseURL)/login") else { throw URLError(.badURL) }
        var req = URLRequest(url: url, timeoutInterval: 12)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.httpBody = try JSONEncoder().encode(["account": account, "password": password])
        let (data, resp) = try await URLSession.shared.data(for: req)
        guard let http = resp as? HTTPURLResponse else { throw URLError(.badServerResponse) }
        guard http.statusCode == 200 else {
            let msg = (try? JSONDecoder().decode([String: String].self, from: data))?["error"]
            throw AuthError.server(msg ?? "登录失败(\(http.statusCode))")
        }
        return try JSONDecoder().decode(AuthResult.self, from: data)
    }

    struct CheckResult: Decodable { let exists: Bool; let hasPassword: Bool }

    /// 统一入口:查邮箱是否已注册、是否设了密码,据此分流。
    @MainActor
    static func check(account: String) async throws -> CheckResult {
        guard let url = URL(string: "\(baseURL)/check") else { throw URLError(.badURL) }
        var req = URLRequest(url: url, timeoutInterval: 12)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.httpBody = try JSONEncoder().encode(["account": account])
        let (data, resp) = try await URLSession.shared.data(for: req)
        guard let http = resp as? HTTPURLResponse else { throw URLError(.badServerResponse) }
        guard http.statusCode == 200 else {
            let msg = (try? JSONDecoder().decode([String: String].self, from: data))?["error"]
            throw AuthError.server(msg ?? "检查失败(\(http.statusCode))")
        }
        return try JSONDecoder().decode(CheckResult.self, from: data)
    }

    /// 验证码登录 - 发码(账号须已注册)。
    @MainActor
    static func loginCode(account: String) async throws {
        guard let url = URL(string: "\(baseURL)/login/code") else { throw URLError(.badURL) }
        var req = URLRequest(url: url, timeoutInterval: 20)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.httpBody = try JSONEncoder().encode(["account": account])
        let (data, resp) = try await URLSession.shared.data(for: req)
        guard let http = resp as? HTTPURLResponse else { throw URLError(.badServerResponse) }
        guard http.statusCode == 200 else {
            let msg = (try? JSONDecoder().decode([String: String].self, from: data))?["error"]
            throw AuthError.server(msg ?? "发送失败(\(http.statusCode))")
        }
    }

    /// 验证码登录 - 验码并登录。
    @MainActor
    static func loginVerify(account: String, code: String) async throws -> AuthResult {
        guard let url = URL(string: "\(baseURL)/login/verify") else { throw URLError(.badURL) }
        var req = URLRequest(url: url, timeoutInterval: 12)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.httpBody = try JSONEncoder().encode(["account": account, "code": code])
        let (data, resp) = try await URLSession.shared.data(for: req)
        guard let http = resp as? HTTPURLResponse else { throw URLError(.badServerResponse) }
        guard http.statusCode == 200 else {
            let msg = (try? JSONDecoder().decode([String: String].self, from: data))?["error"]
            throw AuthError.server(msg ?? "登录失败(\(http.statusCode))")
        }
        return try JSONDecoder().decode(AuthResult.self, from: data)
    }

    /// 注册 - 第一步:请求把注册验证码发到该邮箱(已注册会报错)。
    @MainActor
    static func registerCode(account: String) async throws {
        guard let url = URL(string: "\(baseURL)/register/code") else { throw URLError(.badURL) }
        var req = URLRequest(url: url, timeoutInterval: 20)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.httpBody = try JSONEncoder().encode(["account": account])
        let (data, resp) = try await URLSession.shared.data(for: req)
        guard let http = resp as? HTTPURLResponse else { throw URLError(.badServerResponse) }
        guard http.statusCode == 200 else {
            let msg = (try? JSONDecoder().decode([String: String].self, from: data))?["error"]
            throw AuthError.server(msg ?? "发送失败(\(http.statusCode))")
        }
    }

    /// 注册 - 第二步:验码 + 创建账号,直接返回登录令牌。
    @MainActor
    static func register(account: String, code: String, password: String) async throws -> AuthResult {
        guard let url = URL(string: "\(baseURL)/register") else { throw URLError(.badURL) }
        var req = URLRequest(url: url, timeoutInterval: 12)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.httpBody = try JSONEncoder().encode(["account": account, "code": code, "password": password])
        let (data, resp) = try await URLSession.shared.data(for: req)
        guard let http = resp as? HTTPURLResponse else { throw URLError(.badServerResponse) }
        guard http.statusCode == 200 else {
            let msg = (try? JSONDecoder().decode([String: String].self, from: data))?["error"]
            throw AuthError.server(msg ?? "注册失败(\(http.statusCode))")
        }
        return try JSONDecoder().decode(AuthResult.self, from: data)
    }

    /// 忘记密码 - 第一步:请求把验证码发到该邮箱。
    @MainActor
    static func forgotPassword(account: String) async throws {
        guard let url = URL(string: "\(baseURL)/forgot") else { throw URLError(.badURL) }
        var req = URLRequest(url: url, timeoutInterval: 20)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.httpBody = try JSONEncoder().encode(["account": account])
        let (data, resp) = try await URLSession.shared.data(for: req)
        guard let http = resp as? HTTPURLResponse else { throw URLError(.badServerResponse) }
        guard http.statusCode == 200 else {
            let msg = (try? JSONDecoder().decode([String: String].self, from: data))?["error"]
            throw AuthError.server(msg ?? "发送失败(\(http.statusCode))")
        }
    }

    /// 忘记密码 - 第二步:用验证码重设密码。
    @MainActor
    static func resetPassword(account: String, code: String, password: String) async throws {
        guard let url = URL(string: "\(baseURL)/reset") else { throw URLError(.badURL) }
        var req = URLRequest(url: url, timeoutInterval: 12)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.httpBody = try JSONEncoder().encode(["account": account, "code": code, "password": password])
        let (data, resp) = try await URLSession.shared.data(for: req)
        guard let http = resp as? HTTPURLResponse else { throw URLError(.badServerResponse) }
        guard http.statusCode == 200 else {
            let msg = (try? JSONDecoder().decode([String: String].self, from: data))?["error"]
            throw AuthError.server(msg ?? "重置失败(\(http.statusCode))")
        }
    }

    /// 设置密码:用 Google 登录拿到的 token 给账号设密码,之后可邮箱密码登录。
    @MainActor
    static func setPassword(token: String, password: String) async throws {
        guard let url = URL(string: "\(baseURL)/set-password") else { throw URLError(.badURL) }
        var req = URLRequest(url: url, timeoutInterval: 12)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.httpBody = try JSONEncoder().encode(["token": token, "password": password])
        let (data, resp) = try await URLSession.shared.data(for: req)
        guard let http = resp as? HTTPURLResponse else { throw URLError(.badServerResponse) }
        guard http.statusCode == 200 else {
            let msg = (try? JSONDecoder().decode([String: String].self, from: data))?["error"]
            throw AuthError.server(msg ?? "设置密码失败(\(http.statusCode))")
        }
    }
}

File: cli-anr/app/src/main/java/com/aicodingremote/app/networking/AuthAPI.kt
package com.aicodingremote.app.networking

import com.aicodingremote.app.auth.SessionAuth
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.contentOrNull
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.util.concurrent.TimeUnit

/**
 * 登录 REST(服务器 8080)。对位 iOS `AuthAPI`。
 * Google idToken → 本系统 {account, token};或邮箱密码直登。
 *
 * 端点(沿用设置页服务器 IP,端口固定 8080):
 *   POST /api/auth/login         {account, password}      → AuthResult
 *   POST /api/auth/google        {idToken}                → AuthResult
 *   POST /api/auth/set-password  {token, password}        → 204/200
 *   POST /api/pair/claim         {code, token}            → 200
 */
object AuthAPI {

    /** 登录返回。Google 路径会带 hasPassword:"true"/"false"(是否已设密码)。 */
    @Serializable
    data class AuthResult(
        val account: String,
        val token: String,
        val name: String? = null,
        val hasPassword: String? = null,
    )

    /** 服务端错误(带可展示文案)。 */
    class AuthException(message: String) : Exception(message)

    private val json = Json { ignoreUnknownKeys = true }
    private val jsonMedia = "application/json".toMediaType()

    private val http: OkHttpClient = OkHttpClient.Builder()
        .connectTimeout(12, TimeUnit.SECONDS)
        .readTimeout(12, TimeUnit.SECONDS)
        .build()

    /** REST 基地址:沿用设置页的服务器 IP,端口固定 8080(Spring MVC)。 */
    private fun baseURL(): String = "${ServerConfig.restBaseURL()}/api/auth"

    /** 邮箱密码登录(日常主力:不依赖 Google,只连你自己的服务器)。 */
    suspend fun login(account: String, password: String): AuthResult =
        postForResult(
            "${baseURL()}/login",
            buildJson("account" to account, "password" to password),
            "登录失败",
        )

    /** 用 Google idToken 换本系统会话令牌。 */
    suspend fun loginWithGoogle(idToken: String): AuthResult =
        postForResult(
            "${baseURL()}/google",
            buildJson("idToken" to idToken),
            "登录失败",
        )

    /** 设置密码:用 Google 登录拿到的 token 给账号设密码,之后可邮箱密码登录。 */
    suspend fun setPassword(token: String, password: String) {
        post(
            "${baseURL()}/set-password",
            buildJson("token" to token, "password" to password),
            "设置密码失败",
        )
    }

    /** 认领电脑端的配对码:把 VibeNotch 绑到当前登录账号下。 */
    suspend fun claimPair(code: String) {
        val token = SessionAuth.token ?: throw AuthException("请先登录")
        post(
            "${ServerConfig.restBaseURL()}/api/pair/claim",
            buildJson("code" to code, "token" to token),
            "配对失败",
        )
    }

    // MARK: - 内部

    private suspend fun postForResult(url: String, body: String, failPrefix: String): AuthResult =
        withContext(Dispatchers.IO) {
            val raw = post(url, body, failPrefix)
            json.decodeFromString(AuthResult.serializer(), raw)
        }

    /** 发 POST;非 200 抛带服务端 error 文案的 [AuthException],成功返回响应体文本。 */
    private suspend fun post(url: String, body: String, failPrefix: String): String =
        withContext(Dispatchers.IO) {
            val req = Request.Builder()
                .url(url)
                .post(body.toRequestBody(jsonMedia))
                .build()
            http.newCall(req).execute().use { resp ->
                val text = resp.body?.string().orEmpty()
                if (!resp.isSuccessful) {
                    val msg = parseError(text) ?: "$failPrefix(${resp.code})"
                    throw AuthException(msg)
                }
                text
            }
        }

    private fun parseError(text: String): String? = try {
        (json.parseToJsonElement(text) as? JsonObject)
            ?.get("error")?.let { (it as? JsonPrimitive)?.contentOrNull }
    } catch (_: Throwable) {
        null
    }

    private fun buildJson(vararg pairs: Pair<String, String>): String =
        pairs.joinToString(prefix = "{", postfix = "}", separator = ",") { (k, v) ->
            "\"$k\":\"${escape(v)}\""
        }

    private fun escape(s: String): String =
        s.replace("\\", "\\\\").replace("\"", "\\\"")
}
